csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
Right, but List.AddRange doesn't need to perform any validation - we do. `RepeatedField` prohibits null values because they're not representable in Protocol Buffers, and I view argument validation as a thoroughly good thing.

However, I obviously *do* care about any performance bottlenecks you're seeing. Could you give me more details about your scenario? What is the element type of your repeated field? I'd be willing to ignore (or work out how to handle efficiently) the very niche corner case of value type CLR conversions if it would significantly improve performance. (Heck, you can get LINQ to Objects to break that way, so I think we're in good company.)

For reference types, it's trickier.

- We could validate that the collection didn't contain nulls and then use CopyTo. That wouldn't be thread-safe, but we're not guaranteeing thread safety anyway; I think if the caller is that devious, it's okay for them to only see a problem later.
- We could definitely ensure the right collection size before copying, and inline the Add code so it didn't need to call EnsureSize all the time. I don't know how significant that would be.